clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp

```diff
@@ -41,6 +41,10 @@ AST_MATCHER_P2(Expr, hasSideEffect, bool, CheckFunctionCalls,
   }
 
   if (const auto *OpCallExpr = dyn_cast<CXXOperatorCallExpr>(E)) {
+    if (const auto *MethodDecl =
+            dyn_cast_or_null<CXXMethodDecl>(OpCallExpr->getDirectCallee()))
+      return !MethodDecl->isConst();
```

Comment:
What if called operator is ->, for example in assert we call -> operator on std::shared_ptr just to get member, and then we call const method, this could cause false-positive.

Reply:
I think if the -> operator is implemented correctly, it has to be marked as const, even if it returns something that is not const.

Reply:
On std::shared_ptr yes, but not on std::optional

Reply:
I see